src/box/relay.cc

```diff
@@ -322,6 +327,7 @@ relay_start(struct relay *relay, struct iostream *io, uint64_t sync,
 	relay->io = io;
 	relay->sync = sync;
 	relay->state = RELAY_FOLLOW;
+	relay->sent_raft_term = box_raft()->term;
```

Comment:
The term is sent manually on subscribe, not on final join - `box_process_subscribe()` calls `box_raft_checkpoint_remote()` and then sends the checkpoint together with the term. Also it is sent on initial join - look for `box_raft_checkpoint_local()` in relay.cc.

The reason why it didn't work for you is probably because the initial join has its own send row functions - `relay_send_initial_join_row()`. Also the subscription sends a raft term directly into the file descriptor even before relay is started. So any relay methods won't catch that.

What is more worrying is that I did this in the patch:
```Diff
--- a/src/box/relay.cc
+++ b/src/box/relay.cc
@@ -535,7 +535,6 @@ relay_final_join(struct iostream *io, uint64_t sync,
 		rc = cord_cojoin(&relay->cord);
 	if (rc != 0)
 		diag_raise();
-	relay->sent_raft_term = box_raft()->term;
 
 	ERROR_INJECT(ERRINJ_RELAY_FINAL_JOIN,
```
and all the tests passed. Why?

Reply:
Yeah, this line is not needed. I've checked it with debug prints and it goes something like that:
```log
2022-04-13 07:20:37.610 [3272950] main/146/main I> subscribed replica dfd7a820-65e0-472d-946e-e251b203754b at fd 24, aka 127.0.0.1:3301, peer of 127.0.0.1:51034
2022-04-13 07:20:37.610 [3272950] main/146/main I> remote vclock {1: 8, 2: 5, 3: 2} local vclock {0: 31, 1: 11, 2: 8, 3: 2}
2022-04-13 07:20:37.613 [3272950] relay/127.0.0.1:51034/101/main I> DEBUG relay_raft_msg_push sent 0 req 18
2022-04-13 07:20:37.613 [3272950] relay/127.0.0.1:51034/101/main I> recover from `./00000000000000000000.xlog'
2022-04-13 07:20:37.614 [3272950] relay/127.0.0.1:51034/101/main I> DEBUG relay_send_row sent 18 req 13
2022-04-13 07:20:37.614 [3272950] relay/127.0.0.1:51034/101/main I> DEBUG relay_send_row sent 18 req 14
2022-04-13 07:20:37.614 [3272950] relay/127.0.0.1:51034/101/main I> DEBUG relay_send_row sent 18 req 15
2022-04-13 07:20:37.614 [3272950] relay/127.0.0.1:51034/101/main I> DEBUG relay_send_row sent 18 req 16
2022-04-13 07:20:37.614 [3272950] relay/127.0.0.1:51034/101/main I> DEBUG relay_send_row sent 18 req 17
2022-04-13 07:20:37.614 [3272950] relay/127.0.0.1:51034/101/main I> DEBUG relay_send_row sent 18 req 18
```

Reply:
Now I can answer here, so I'm pasting a comment from my review:

It might happen that the new term is sent right on box_process_subscribe, outside of relay_subscribe loop. And we forgot to set relay->sent_raft_term there.
I didn't want to access relay directly from box, so I've set sent_raft_term to current value in relay_start. By the time relay_start() is called, the term is either sent (during subscribe) or the term doesn't matter (during join).

We can just set sent_raft_term to current value in box_process_subscribe(), but it doesn't look too clean IMO.

Reply:
On further investigation I get the point. There is a direct coio_write_xrow call, sending current RAFT state to replica in box_process_subscribe. How about following solution.

```
diff --git a/src/box/relay.cc b/src/box/relay.cc
index 93b5b571a..fcdcc09da 100644
--- a/src/box/relay.cc
+++ b/src/box/relay.cc
@@ -474,6 +474,7 @@ relay_initial_join(struct iostream *io, uint64_t sync, struct vclock *vclock,

                xrow_encode_raft(&row, &fiber()->gc, &raft_req);
                xstream_write(&relay->stream, &row);
+               relay->sent_raft_term = raft_req.term;

                char body[XROW_SYNCHRO_BODY_LEN_MAX];
                xrow_encode_synchro(&row, body, &req);
@@ -1019,6 +1020,12 @@ relay_subscribe(struct replica *replica, struct iostream *io, uint64_t sync,

        relay->id_filter = replica_id_filter;

+       /**
+        * This term was just sent in box_process_subscribe directly via
+        * coio_write_xrow call
+        */
+       relay->sent_raft_term = box_raft()->term;
+
        int rc = cord_costart(&relay->cord, "subscribe",
                              relay_subscribe_f, relay);
        if (rc == 0)
```

Reply:
It might work. But I don't understand why it works without this diff too. It is not needed if it can't be proven to change anything.

Reply:
> How about following solution.

@grafin, sounds good.

> But I don't understand why it works without this diff too

@Gerold103 me neither. I remember adding this line to fix some of the flaky hangs in tests. I couldn't reproduce the hangs now, though.

Reply:
@grafin, thanks for detailed investigation!
We are good to go then.